nums/core/linalg.py

```diff
@@ -256,6 +256,147 @@ def inv(app: ArrayApplication, X: BlockArray):
     return result
 
 
+def inv_uppertri(app: ArrayApplication, X: BlockArray):
+    # Inversion of an Upper Triangular Matrix
+    # Use the method described in https://www.cs.utexas.edu/users/flame/pubs/siam_spd.pdf
+    assert len(X.shape) == 2
+    assert X.shape[0] == X.shape[1], "This function only accepts square matrices"
+    single_block = X.shape[0] == X.block_shape[0] and X.shape[1] == X.block_shape[1]
+    nonsquare_block = X.block_shape[0] != X.block_shape[1]
+
+    # If X is single block or block size is non-square, then reshape BS to row_size // 4
+    if single_block or nonsquare_block:
```

Comment:
If X is a single block, why not return `inv`?

Reply:
Hmm, I see. Let's keep it the way it is for now. `inv` will be fast, but single-threaded. There might be a point beyond which what you're proposing will be faster, but we would need to test and apply heuristics accordingly (e.g. if the array is big then partition and use optimized imp).